ntp-client/src/Network/NTP/Client.hs

```diff
@@ -71,248 +64,221 @@ data NtpStatus =
-        if null addrs
-          then do
-            atomically $ writeTVar ncStatus NtpSyncUnavailable
-            -- either wait 30s or wait for `NtpSyncPending` which might be set
-            -- by a client (e.g. wallet), in which case try to resolve the dns.
-            race_
-              (threadDelay 30000000)
-              (do
-                atomically $ do
-                  s <- readTVar ncStatus
-                  case s of
-                    NtpSyncPending -> return ()
-                    _              -> retry
-                traceWith tracer NtpTraceSpawnNtpClientResolvePending
-              )
-            resolve
-          else return addrs
-
--- Try to create IPv4 and IPv6 socket.
-mkSockets :: Tracer IO NtpTrace -> NtpClientSettings -> IO Sockets
-mkSockets tracer settings =
-    doMkSockets `catch` handleIOException >>= \case
-        Just socks -> pure socks
-        Nothing     -> do
-            traceWith tracer NtpTraceMkSocketsNoSockets
---            logWarning "Couldn't create either IPv4 or IPv6 socket, retrying in 5 sec..."
-            threadDelay 5000000
-            mkSockets tracer settings
+        runProtocol _version _localAddr [] = return []
+        runProtocol _version Nothing    _  = return []
+        runProtocol version (Just addr) servers = do
+            socketAction tracer ntpSettings addr servers >>= \case
+                Left err -> do
+                    traceWith tracer $ NtpTraceRunProtocolError version err
+                    return []
+                Right [] -> do
+                    traceWith tracer $ NtpTraceRunProtocolNoResult version
+                    return []
+                Right r@(_:_) -> do
+                    traceWith tracer $ NtpTraceRunProtocolSuccess version
+                    return r
+
+socketAction ::
+       Tracer IO NtpTrace
+    -> NtpClientSettings
+    -> AddrInfo
+    -> [AddrInfo]
+    -> IO (Either IOError [ReceivedPacket])
+socketAction tracer netSettings localAddr destAddrs
+    = tryIOError $ bracket acquire release action
   where
-    doMkSockets :: IO (Maybe Sockets)
-    doMkSockets = do
-        addrs <- udpLocalAddresses
-        (<>) <$> (createAndBindSock tracer IPv4 addrs)
-             <*> (createAndBindSock tracer IPv6 addrs)
-
-    handleIOException :: IOException -> IO (Maybe Sockets)
-    handleIOException e = do
-        traceWith tracer $ NtpTraceMkSocketsIOExecption e
-        threadDelay 5000000
-        doMkSockets
+    acquire :: IO Socket
+    acquire = do
+        s <- Socket.socket (addrFamily localAddr) Datagram Socket.defaultProtocol
+        traceWith tracer NtpTraceSocketOpen
+        return s
+
+    release :: Socket -> IO ()
+    release s = do
+        Socket.close s
+        traceWith tracer NtpTraceSocketClosed
+
+    action :: Socket -> IO [ReceivedPacket]
+    action socket = do
+        Socket.setSocketOption socket ReuseAddr 1
+        Socket.bind socket (addrAddress localAddr)
```

Comment:
if we don't bind, we will get an ephemeral port, which will be different each time we run the loop. This will make it even harder to DDOS.

Reply:
Interesting. I didn't know about that. I was just following the pattern of the old ntp-client.